File: aoc2017/15.py
QUIZ_NUMBER = "15"

class Solution:
    def run(fileName = QUIZ_NUMBER + ".in"):
        solution = Solution(fileName)
        solution.solve1()
        solution.solve2()

    def test():
        Solution.run(QUIZ_NUMBER + ".ex.in")

    def __init__(self, fileName):
        self.ins = []
        self.fileName = fileName
        fp = open(fileName, 'r')
        self.ins = list(map(int, fp.read().split()))
        fp.close()
        print("===", self.fileName, "===")

    def solve1(self):
        print("--- Part One ---")
        a = self.ins[0]
        b = self.ins[1]
        af = 16807
        bf = 48271
        d = 2147483647
        finalCount = 0
        for _ in range(40000000):
            a *= af
            a %= d
            b *= bf
            b %= d
            ar = a & 0xFFFF
            br = b & 0xFFFF
            if ar == br:
                finalCount += 1
        print(f"Judge's final count: {finalCount}")

    def solve2(self):
        print("--- Part Two ---")

def main():
    Solution.test()
    Solution.run()

if __name__ == "__main__":
    main()

File: aoc2017/15.cpp
#include <queue>
#include <iostream>
#include <string>
#include <unordered_map>
#include <cstdint>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;
using ull = uint64_t;

void part1(ull a, ull b) {
    cout << "--- Part 1 ---" << endl;
    constexpr const ull af = 16807;
    constexpr const ull bf = 48271;
    constexpr const ull d = 2147483647;
    ull finalCount = 0;
    for (ull i = 0; i < 40000000; ++i) {
        a *= af;
        a %= d;
        b *= bf;
        b %= d;
        ull ar = a & 0xFFFF;
        ull br = b & 0xFFFF;
        if (ar == br)
            finalCount += 1;
    }
    cout << "Judge's final count: " << finalCount << endl;
}

void part2(ull a, ull b) {
    cout << "--- Part 2 ---" << endl;
    constexpr const ull af = 16807;
    constexpr const ull bf = 48271;
    constexpr const ull d = 2147483647;
    ull finalCount = 0;
    for (ull i = 0; i < 5000000; ++i) {
        do {
            a *= af;
            a %= d;
        } while (a % 4 > 0);
        do {
            b *= bf;
            b %= d;
        } while (b % 8 > 0);
        ull ar = a & 0xFFFF;
        ull br = b & 0xFFFF;
        if (ar == br)
            finalCount += 1;
    }
    cout << "Judge's final count: " << finalCount << endl;
}

int main(void) {
    cout << "== 15.ex.in ==" << endl;
    part1(65,8921);
    part2(65,8921);
    cout << "== 15.in ==" << endl;
    part1(873,583);
    part2(873,583);
    return 0;
}